google-ads-examples/src/main/java/com/google/ads/googleads/examples/shoppingads/AddPerformanceMaxProductListingGroupTree.java

```diff
@@ -0,0 +1,554 @@
+        HashSet<String> siblings;
+        // Check to see if we've already visited a sibling in this group, and fetch or create a new
+        // set as required.
+        if (this.parentsToChildren.containsKey(parentResourceName)) {
+          siblings = this.parentsToChildren.get(parentResourceName);
+        } else {
+          siblings = new HashSet<>();
+        }
+
+        siblings.add(filter.getResourceName());
+        this.parentsToChildren.put(parentResourceName, siblings);
+      }
+    }
+
+    // [START add_performance_max_product_listing_group_tree_2]
+    /**
+     * Creates a list of MutateOperations that remove all of the resources in the tree originally
+     * used to create this factory object.
+     */
+    public List<MutateOperation> removeAll() {
+      return removeDescendantsAndFilter(rootResourceName);
+    }
+    // [END add_performance_max_product_listing_group_tree_2]
+
+    // [START add_performance_max_product_listing_group_tree_3]
+    /**
+     * Creates a list of MutateOperations that remove all the descendants of the specified
+     * AssetGroupListingGroupFilter resource name. The order of removal is post-order, where all the
+     * children (and their children, recursively) are removed first. Then, the node itself is
+     * removed.
+     */
+    public List<MutateOperation> removeDescendantsAndFilter(String resourceName) {
+      List<MutateOperation> operations = new ArrayList<>();
+
+      if (this.parentsToChildren.containsKey(resourceName)) {
+        HashSet<String> children = parentsToChildren.get(resourceName);
+        for (String child : children) {
+          operations.addAll(removeDescendantsAndFilter(child));
+        }
+      }
+
+      AssetGroupListingGroupFilterOperation operation =
+          AssetGroupListingGroupFilterOperation.newBuilder().setRemove(resourceName).build();
+      operations.add(
+          MutateOperation.newBuilder().setAssetGroupListingGroupFilterOperation(operation).build());
+      return operations;
+    }
+  }
+  // [END add_performance_max_product_listing_group_tree_3]
+
+  /**
+   * A factory that creates MutateOperations wrapping AssetGroupListingGroupFilterMutateOperations
+   * for a specific customerId and assetGroupId.
+   *
+   * <p>This object is intended to be used with an array of MutateOperations to perform an atomic
+   * update to an AssetGroup.
+   */
+  private static class AssetGroupListingGroupFilterCreateOperationFactory {
+    private final long customerId;
+    private final long assetGroupId;
+    private final long rootListingGroupId;
+    private long nextId;
+
+    public AssetGroupListingGroupFilterCreateOperationFactory(
+        long customerId, long assetGroupId, long rootListingGroupId) {
+      this.customerId = customerId;
+      this.assetGroupId = assetGroupId;
+      this.rootListingGroupId = rootListingGroupId;
+      this.nextId = this.rootListingGroupId;
```

Comment:
Up to you, but you could replace this with an iterator from a [`LongStream`](https://docs.oracle.com/javase/9/docs/api/java/util/stream/LongStream.html).